### PR TITLE
Migrate tests to jest

### PR DESCRIPTION
**Pull request by asantos00**
- Starts migrating tests to jest

I've built a rough (very ugly and bad, please don't cry while reading it) to do most of the find and replaces. I'll leave it below, as it might help.

I'll keep it updated as it also works as a "recipe".

```javascript
const path = require('path');
const replaceInFile = require('replace-in-file')

const [file, repoRoot] = process.argv.slice(2);

(async () => {
  try {
    console.log('Replace assert');
    await replaceInFile({
      files: file,
      from: /\(assert\) => {/g,
      to: `() => {`
    });
    await replaceInFile({
      files: file,
      from: /assert => {/g,
      to: `() => {`
    });

    console.log('Replace assert - old way');
    await replaceInFile({
      files: file,
      from: /function\(assert\) {/g,
      to: `function() {`
    });

    console.log('Replace module');
    await replaceInFile({
      files: file,
      from: /module\(/g,
      to: `describe(`
    });

    console.log('Remove qunits')
    await replaceInFile({
      files: file,
      from: /import { module, test } from 'qunit';\n/g,
      to: ``
    });
    await replaceInFile({
      files: file,
      from: /import {module, test} from 'qunit';\n/g,
      to: ``
    });
    await replaceInFile({
      files: file,
      from: /import { module } from 'qunit';\n/g,
      to: ``
    });

    console.log('Fix path to @miragejs/server')
    const fileFolder = file.split('/')
    fileFolder.pop();
    await replaceInFile({
      files: file,
      from: /'@miragejs\/server'/g,
      to: `'${path.relative(fileFolder.join('/'), repoRoot)}/lib'`
    });

    console.log('Fix assert.expect')
    await replaceInFile({
      files: file,
      from: /assert\.expect\(/g,
      to: `expect.assertions(`
    });

    console.log('Fix hooks')
    await replaceInFile({
      files: file,
      from: /hooks\./g,
      to: ``
    });
    await replaceInFile({
      files: file,
      from: /function\(hooks\)/g,
      to: `function()`
    });
  } catch(e) {
    console.log('Error in replaces', e)
  }

})()
```

**Usage**: `node migrate-to-jest [path to test] [path to @miragejs/server repo root]`.

## Things that need to be migrated manually

- When tests use context (`this.server` for instance)

### CODE REVIEW

File: __tests__/unit/collection-test.js [hunk 1]
```diff
@@ -1,47 +1,46 @@
...
-    assert.ok(collection);
+    expect(collection).toBeTruthy();
   });
 
-  test('it cannot be instantiated without a modelName', function(assert) {
-    assert.throws(() => {
+  test('it cannot be instantiated without a modelName', () => {
+    expect(() => {
       new Collection();
-    }, /must pass a `modelName`/);
```
Comment (samselikoff) [inline __tests__/unit/collection-test.js]: looks like there's a `toThrowError(/must pass a `modelName`/)` we can use here. Would be good to preserve these error message checks.

File: __tests__/unit/model-test.js [hunk 2]
```diff
@@ -1,32 +1,31 @@
...
-    assert.ok(model);
+    expect(model).toBeTruthy();
   });
 
-  test('it cannot be instantiated without a schema', function(assert) {
-    assert.throws(function() {
+  test('it cannot be instantiated without a schema', assert => {
+    expect(function() {
       new Model();
-    }, /requires a schema/);
```
Comment (samselikoff) [inline __tests__/unit/model-test.js]: same

File: __tests__/unit/model-test.js [hunk 3]
```diff
@@ -1,32 +1,31 @@
...
+    }).toThrow();
   });
 
-  test('it cannot be instantiated without a modelName', function(assert) {
-    assert.throws(function() {
+  test('it cannot be instantiated without a modelName', assert => {
+    expect(function() {
       new Model({});
-    }, /requires a modelName/);
```
Comment (samselikoff) [inline __tests__/unit/model-test.js]: same

File: __tests__/unit/reference-sort-test.js [hunk 4]
```diff
@@ -1,38 +1,35 @@
...
+  test('it throws on circular dependency', () => {
+    expect(function() {
       referenceSort([
         ['propA', 'propB'],
         ['propB', 'propA']
       ]);
-    }, function(e) {
-      return e.toString() === 'Error: Cyclic dependency in properties ["propB","propA"]';
-    });
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/reference-sort-test.js]: same

File: __tests__/unit/server-test.js [hunk 5]
```diff
@@ -82,39 +81,39 @@ module('Unit | Server #db', function() {
...
       factories: {
         address: Factory
       }
     });
 
-    assert.throws(function() {
+    expect(function() {
       server.create('contact');
-    }, /no model or factory was found/);
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 6]
```diff
@@ -474,14 +477,14 @@ module('Unit | Server #create', function() {
       }
     });
 
-    assert.throws(() => {
+    expect(() => {
       server.create('article', 'private');
-    }, /'private' trait is not registered in 'article' factory/);
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 7]
```diff
@@ -744,16 +747,16 @@ module('Unit | Server #createList', function(hooks) {
...
+    server.loadFactories({
       article: ArticleFactory
     });
 
-    assert.throws(() => {
-      this.server.createList('article', 2, 'private');
-    }, /'private' trait is not registered in 'article' factory/);
+    expect(() => {
+      server.createList('article', 2, 'private');
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 8]
```diff
@@ -763,59 +766,60 @@ module('Unit | Server #createList', function(hooks) {
...
     });
 
-    this.server.loadFactories({
+    server.loadFactories({
       article: ArticleFactory
     });
 
-    assert.throws(() => {
-      this.server.createList('article', 'published');
-    }, /second argument has to be an integer, you passed: string/);
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 9]
```diff
@@ -1100,13 +1103,13 @@ module('Unit | Server #build', function(hooks) {
       }
     });
 
-    assert.throws(() => {
-      this.server.build('article', 'withCategory');
-    }, /Model not registered: article/);
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 10]
```diff
@@ -1129,77 +1132,78 @@ module('Unit | Server #build', function(hooks) {
       }
     });
 
-    assert.throws(() => {
-      this.server.build('article', 'withCategory');
-    }, /You're using the `association` factory helper on the 'category' attribute of your article factory/);
+    expect(() => {
+      server.build('article', 'withCategory');
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 11]
```diff
@@ -1266,16 +1270,16 @@ module('Unit | Server #buildList', function(hooks) {
...
+    server.loadFactories({
       article: ArticleFactory
     });
 
-    assert.throws(() => {
-      this.server.buildList('article', 2, 'private');
-    }, /'private' trait is not registered in 'article' factory/);
+    expect(() => {
+      server.buildList('article', 2, 'private');
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/server-test.js [hunk 12]
```diff
@@ -1285,40 +1289,40 @@ module('Unit | Server #buildList', function(hooks) {
...
+    server.loadFactories({
       article: ArticleFactory
     });
 
-    assert.throws(() => {
-      this.server.buildList('article', 'published');
-    }, /second argument has to be an integer, you passed: string/);
+    expect(() => {
+      server.buildList('article', 'published');
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/server-test.js]: same

File: __tests__/unit/db/identity-manager-test.js [hunk 13]
```diff
@@ -1,86 +1,86 @@
...
-  test(`set indicates an id is being used`, function(assert) {
+  test(`set indicates an id is being used`, () => {
     let manager = new IdentityManager();
     manager.set('abc');
 
-    assert.throws(function() {
+    expect(function() {
       manager.set('abc');
-    }, /already been used/);
+    }).toThrow();
```
Comment (samselikoff) [inline __tests__/unit/db/identity-manager-test.js]: same

**Comment by samselikoff** [comment]
wow man, amazing work!!

Just a few spots where it'd be nice to keep error messages in expect().toThrow.

After that should we merge these?

**Comment by asantos00** [comment]
You're definitely right, I'll do that fix. I think so, we should keep merging it so we don't loose track of the PR and it does not get too big 👍

**Pull request merged**